From d73ca7282ce93f70334950640d01d7071c666d3e Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Mon, 22 Sep 2014 19:41:36 -0400
Subject: [crypto] pkcs7: Provide a single place to do signed info block freeing

Message-id: <1411414897-26220-4-git-send-email-vgoyal@redhat.com>
Patchwork-id: 96446
O-Subject: [RHEL7.1 PATCH 3/4] PKCS#7: Provide a single place to do signed info block freeing
Bugzilla: 1129779
RH-Acked-by: David Howells <dhowells@redhat.com>

BZ:1129779

This commit id is in linux-security tree of James Morris.

commit 3cd0920cde6a726b9965b60184f9d5a54358304d
Author: David Howells <dhowells@redhat.com>
Date: Tue, 16 Sep 2014 17:29:03 +0100

The code to free a signed info block is repeated several times, so move the
code to do it into a function of its own.  This gives us a place to add clean
ups for stuff that gets added to pkcs7_signed_info.

Signed-off-by: David Howells <dhowells@redhat.com>
Acked-by: Vivek Goyal <vgoyal@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/asymmetric_keys/pkcs7_parser.c b/crypto/asymmetric_keys/pkcs7_parser.c
index 42e56aa..4c4ea35 100644
--- a/crypto/asymmetric_keys/pkcs7_parser.c
+++ b/crypto/asymmetric_keys/pkcs7_parser.c
@@ -31,6 +31,18 @@ struct pkcs7_parse_context {
  unsigned sinfo_index;
 };
 
+/*
+ * Free a signed information block.
+ */
+static void pkcs7_free_signed_info(struct pkcs7_signed_info *sinfo)
+{
+ if (sinfo) {
+  mpi_free(sinfo->sig.mpi[0]);
+  kfree(sinfo->sig.digest);
+  kfree(sinfo);
+ }
+}
+
 /**
  * pkcs7_free_message - Free a PKCS#7 message
  * @pkcs7: The PKCS#7 message to free
@@ -54,9 +66,7 @@ void pkcs7_free_message(struct pkcs7_message *pkcs7)
   while (pkcs7->signed_infos) {
    sinfo = pkcs7->signed_infos;
    pkcs7->signed_infos = sinfo->next;
-   mpi_free(sinfo->sig.mpi[0]);
-   kfree(sinfo->sig.digest);
-   kfree(sinfo);
+   pkcs7_free_signed_info(sinfo);
   }
   kfree(pkcs7);
  }
@@ -100,16 +110,12 @@ struct pkcs7_message *pkcs7_parse_message(const void *data, size_t datalen)
   ctx->certs = cert->next;
   x509_free_certificate(cert);
  }
- mpi_free(ctx->sinfo->sig.mpi[0]);
- kfree(ctx->sinfo->sig.digest);
- kfree(ctx->sinfo);
+ pkcs7_free_signed_info(ctx->sinfo);
  kfree(ctx);
  return msg;
 
 error_decode:
- mpi_free(ctx->sinfo->sig.mpi[0]);
- kfree(ctx->sinfo->sig.digest);
- kfree(ctx->sinfo);
+ pkcs7_free_signed_info(ctx->sinfo);
 error_no_sinfo:
  kfree(ctx);
 error_no_ctx:
-- 
1.7.1